frontend/lib/resolution/prims.cpp

```diff
@@ -740,11 +740,29 @@ primComplexGetComponent(Context* context, const CallInfo& ci) {
   if (ci.numActuals() != 1) return ret;
 
   if (auto comp = ci.actual(0).type().type()->toComplexType()) {
-    if (comp->bitwidth() == 64) {
-      ret = QualifiedType(QualifiedType::REF, RealType::get(context, 32));
-    } else if (comp->bitwidth() == 128) {
-      ret = QualifiedType(QualifiedType::REF, RealType::get(context, 64));
-    }
+    int w = comp->componentBitwidth();
+    ret = QualifiedType(QualifiedType::REF, RealType::get(context, w));
+  }
+  return ret;
+}
+
+/* for abs */
+static QualifiedType
+primAbsGetType(Context* context, const CallInfo& ci) {
+  QualifiedType ret = QualifiedType();
+
+  if (ci.numActuals() != 1) return ret;
+
+  QualifiedType actualType = ci.actual(0).type();
+  if (auto comp = actualType.type()->toComplexType()) {
+    int w = comp->componentBitwidth();
+    ret = QualifiedType(QualifiedType::CONST_VAR, RealType::get(context, w));
+  } else if (auto img = actualType.type()->toImagType()) {
+    int w = img->bitwidth();
+    ret = QualifiedType(QualifiedType::CONST_VAR, RealType::get(context, w));
+  } else {
+    // otherwise just use the original type
+    ret = QualifiedType(QualifiedType::CONST_VAR, actualType.type());
```

Comment:
Note that the production compiler treats ABS and SQRT as param-only primitives. That leads me to think that Dyno should also implement the param logic, and not support the non-param case.

Reply:
OK, I added some comments along these lines, and will note it in Future Work.